sdks/python/apache_beam/runners/interactive/caching/file_based_cache.py

```diff
@@ -0,0 +1,311 @@
+      writer_kwargs (Dict[str, Any]): A dictionary of key-value pairs
+          to be passed to the underlying writer objects.
+
+    Raises:
+      ~exceptions.ValueError: If the specified mode is not supported.
+      ~exceptions.IOError: If one or more files matching `location` already
+          exist and mode == "error".
+    """
+    self.location = location
+    self.element_type = None
+    self.default_coder = writer_kwargs.pop("coder", None)
+    self._writer_kwargs = writer_kwargs
+    self._num_writes = 0
+    self._persist = persist
+    self._finalizer = self._configure_finalizer(persist)
+
+    # TODO(ostrokach): Implement append mode.
+    if mode not in ['error', 'overwrite']:
+      raise ValueError("'mode' must be set to one of: ['error', 'overwrite'].")
+    exitsting_files = list(glob_files(self.file_pattern))
+    if mode == "error" and exitsting_files:
+      raise IOError("The following cache files already exist: {}.".format(
+          exitsting_files))
+    if mode == "overwrite":
+      self.truncate()
+
+    root, _ = FileSystems.split(self._file_path_prefix)
+    try:
+      FileSystems.mkdirs(root)
+    except IOError:
+      pass
+    # It is possible to read from am empty stream, so it should also be possible
+    # to read from an empty file.
+    FileSystems.create(self._file_path_prefix + ".empty").close()
+
+  @property
+  def persist(self):
+    return self._persist
+
+  @persist.setter
+  def persist(self, persist):
+    if self._persist == persist:
+      return
+    self._persist = persist
+    if self._finalizer:
+      self._finalizer.detach()
+    self._finalizer = self._configure_finalizer(persist)
+
+  @property
+  def timestamp(self):
+    timestamp = 0
+    for path in glob_files(self.file_pattern):
+      timestamp = max(timestamp, FileSystems.last_updated(path))
+    return timestamp
+
+  def reader(self, **kwargs):
+    """Returns a reader ``PTransform`` to be used for reading the contents of
+    the cache into a Beam Pipeline.
+
+    Args:
+      kwargs (Dict[str, Any]): Arguments to be passed to the underlying reader
+          class.
+    """
+    reader_kwargs = self._reader_kwargs.copy()
+    reader_kwargs.update(kwargs)
+    reader = self._reader_class(self.file_pattern, **reader_kwargs)
+    # Keep a reference to the parent object so that cache does not get garbage
+    # collected while the pipeline is running.
+    reader._cache = self
+    return reader
+
+  def writer(self):
+    """Returns a writer object to be used for writing the contents of a
+    PCollection from a Beam Pipeline into the cache.
+    """
+    self._num_writes += 1
+    writer_kwargs = self._writer_kwargs.copy()
+
+    if self.element_type is None:
+      writer = PatchedWriter(self, self._writer_class,
+                             (self._file_path_prefix,), writer_kwargs)
+      return writer
+
+    if ("coder" in self._reader_passthrough_arguments and
```

Comment:
What defines `_reader_passthrough_arguments`? All the subclasses?

Reply:
The idea behind `_reader_passthrough_arguments` was to specify the set of `writer_kwargs` that should also be passed onto the reader PTransform (such as `coder`, `compression_type`, etc.).

Since this attribute has come up in several of your comments, I understand that it is confusing to the user and probably not something that they should worry about. Therefore, I hard-coded this parameter in the specific cache implementations (e.g. see the `__init__` methods of `TextBasdCache`, `TFRecordBasedCace`, etc.).